contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,1098 @@
+    string public constant REBASE_CHECK_NEAREST_EPOCH_DISTANCE_KEY = "REBASE_CHECK_NEAREST_EPOCH_DISTANCE";
+    bytes public constant REBASE_CHECK_NEAREST_EPOCH_DISTANCE_VALUE = hex"04";
+    string public constant REBASE_CHECK_DISTANT_EPOCH_DISTANCE_KEY = "REBASE_CHECK_DISTANT_EPOCH_DISTANCE";
+    bytes public constant REBASE_CHECK_DISTANT_EPOCH_DISTANCE_VALUE = hex"0a";
+    string public constant VALIDATOR_DELAYED_TIMEOUT_IN_SLOTS_KEY = "VALIDATOR_DELAYED_TIMEOUT_IN_SLOTS";
+    bytes public constant VALIDATOR_DELAYED_TIMEOUT_IN_SLOTS_VALUE = hex"1c20";
+    string public constant VALIDATOR_DELINQUENT_TIMEOUT_IN_SLOTS_KEY = "VALIDATOR_DELINQUENT_TIMEOUT_IN_SLOTS";
+    bytes public constant VALIDATOR_DELINQUENT_TIMEOUT_IN_SLOTS_VALUE = hex"5460";
+    string public constant PREDICTION_DURATION_IN_SLOTS_KEY = "PREDICTION_DURATION_IN_SLOTS";
+    bytes public constant PREDICTION_DURATION_IN_SLOTS_VALUE = hex"c4e0";
+    string public constant FINALIZATION_MAX_NEGATIVE_REBASE_EPOCH_SHIFT_KEY = "FINALIZATION_MAX_NEGATIVE_REBASE_EPOCH_SHIFT";
+    bytes public constant FINALIZATION_MAX_NEGATIVE_REBASE_EPOCH_SHIFT_VALUE = hex"0546";
+    string public constant NODE_OPERATOR_NETWORK_PENETRATION_THRESHOLD_BP_KEY = "NODE_OPERATOR_NETWORK_PENETRATION_THRESHOLD_BP";
+    bytes public constant NODE_OPERATOR_NETWORK_PENETRATION_THRESHOLD_BP_VALUE = hex"64";
+
+    //
+    // Immutables
+    //
+    // Timestamp since startUpgrade() and finishUpgrade() revert with Expired()
+    // This behavior is introduced to disarm the template if the upgrade voting creation or enactment didn't
+    // happen in proper time period
+    uint256 immutable public EXPIRE_SINCE_INCLUSIVE;
+
+    //
+    // Structured storage
+    //
+    /// UPGRADE_NOT_STARTED (zero) by default
+    uint256 public _upgradeBlockNumber;
+    bool public _isUpgradeFinished;
+
+    constructor(uint256 expireSinceInclusive) {
+        if (expireSinceInclusive <= block.timestamp) revert ExpireSinceMustBeInFuture();
+
+        EXPIRE_SINCE_INCLUSIVE = expireSinceInclusive;
+
+        emit TemplateCreated(EXPIRE_SINCE_INCLUSIVE);
+    }
+
+    /// @notice Need to be called before LidoOracle implementation is upgraded to LegacyOracle
+    function startUpgrade() external {
+        _assertNotExpired();
+        _startUpgrade();
+        emit UpgradeStarted();
+    }
+
+    /// @notice Need to be called after LidoOracle implementation is upgraded to LegacyOracle
+    function finishUpgrade() external {
+        _assertNotExpired();
+        _finishUpgrade();
+        emit UpgradeFinished();
+    }
+
+    /// @notice Perform basic checks to revert the entire upgrade if something gone wrong
+    function assertUpgradeIsFinishedCorrectly() external view {
+        _assertUpgradeIsFinishedCorrectly();
+    }
+
+    /// @notice Needed for 2nd Aragon voting (roles revoke) to fail if 1st voting isn't enacted
+    function revertIfUpgradeNotEnacted() external view {
```

Comment:
remove duplicate of the check and just make a single public function and single error